events/consumers.go

```diff
@@ -0,0 +1,196 @@
+		logger.Errorf("Failed to process event[%s]: %s", e.ID, e.Error)
+	}
+
+	if err := tx.Create(failedEvents).Error; err != nil {
+		// TODO: More robust way to handle failed event insertion failures
+		logger.Errorf("Error inserting into table:event_queue with error:%v. %v", err)
+	}
+	return tx.Commit().Error
+}
+
+// ConsumeEventsUntilEmpty consumes events forever until the event queue is empty.
+func (t *EventConsumer) ConsumeEventsUntilEmpty() {
+	consumerFunc := func(wg *sync.WaitGroup) {
+		for {
+			err := t.consumeEvents()
+			if err != nil {
+				if errors.Is(err, gorm.ErrRecordNotFound) {
+					wg.Done()
+					return
+				} else {
+					logger.Errorf("error processing event, waiting 60s to try again %v", err)
+					time.Sleep(waitDurationOnFailure)
+				}
+			}
+		}
+	}
+
+	var wg sync.WaitGroup
+	for i := 0; i < t.Consumers; i++ {
+		wg.Add(1)
+		go consumerFunc(&wg)
+	}
+	wg.Wait()
+}
+
+// listenToPostgresNotify listens to postgres notifications
+// and will retry on failure for dbReconnectMaxAttempt times
+func (e *EventConsumer) listenToPostgresNotify(pgNotify chan bool) {
+	var listen = func(ctx context.Context, pgNotify chan bool) error {
+		conn, err := db.Pool.Acquire(ctx)
+		if err != nil {
+			return fmt.Errorf("error acquiring database connection: %v", err)
+		}
+		defer conn.Release()
+
+		_, err = conn.Exec(ctx, "LISTEN event_queue_updates")
+		if err != nil {
+			return fmt.Errorf("error listening to database notifications: %v", err)
+		}
+		logger.Infof("listening to database notifications")
+
+		for {
+			notification, err := conn.Conn().WaitForNotification(ctx)
+			if err != nil {
+				return fmt.Errorf("error listening to database notifications: %v", err)
+			}
+
+			logger.Tracef("Received database notification: %+v", notification)
+			pgNotify <- true
+		}
+	}
+
+	// retry on failure.
+	for {
+		backoff := retry.WithMaxDuration(dbReconnectMaxDuration, retry.NewExponential(dbReconnectBackoffBaseDuration))
+		err := retry.Do(context.TODO(), backoff, func(ctx context.Context) error {
+			if err := listen(ctx, pgNotify); err != nil {
+				return retry.RetryableError(err)
+			}
+
+			return nil
+		})
+
+		logger.Errorf("failed to connect to database: %v", err)
+	}
+}
+
+func (e *EventConsumer) Listen() {
+	logger.Infof("Started listening for database notify events")
+
+	if err := e.Validate(); err != nil {
+		logger.Fatalf("Error starting event consumer: %v", err)
+		return
+	}
+
+	// Consume pending events
+	e.ConsumeEventsUntilEmpty()
+
+	pgNotify := make(chan bool)
+	go e.listenToPostgresNotify(pgNotify)
+
+	for {
+		select {
+		case <-pgNotify:
+			e.ConsumeEventsUntilEmpty()
+
+		case <-time.After(pgNotifyTimeout):
+			e.ConsumeEventsUntilEmpty()
+		}
+	}
+}
+
+func StartConsumers(gormDB *gorm.DB, config Config) {
+	allConsumers := []EventConsumer{
+		NewTeamConsumer(gormDB),
+		NewNotificationConsumer(gormDB),
+		NewResponderConsumer(gormDB),
+		NewUpstreamPushConsumer(gormDB, config),
```

Comment:
I think we should not be setting up consumer for upstream if the upstream config is not fully provided.
Right now it panics [here](https://github.com/flanksource/incident-commander/pull/458/files#diff-5f3bc2bbe3b2795d2d392d7d394a13f7c4d422de3ae1a45fc7371d363cfa6402R29)

Reply:
That was intentional, it will panic if we do not setup the agent mode but still receive push events.

Should I change it to not creating one ?

Reply:
Have updated 👍🏽